### PR TITLE
Redesign of UCXAddress and implementation of ep_create_from_address() 

### PR DESCRIPTION
**Pull request by madsbk**
Minimize the use of pure C

This PR is part **3** of an effort to merge https://github.com/rapidsai/ucx-py/pull/648 in a divide-and-conquer approach.

### CODE REVIEW

**Comment by pentschev** [comment]
@madsbk failures are related to this PR, perhaps it's only a matter of updating `test_send_recv_addr` ?

**Comment by madsbk** [comment]
> @madsbk failures are related to this PR, perhaps it's only a matter of updating `test_send_recv_addr` ?

Yes, I am rewriting `test_send_recv_addr` as well. Will let you know when I have something ready for review :)

File: ucp/_libs/ucx_api.pyx [hunk 1]
```diff
@@ -10,12 +10,17 @@ import weakref
 
 from posix.stdio cimport open_memstream
 
-from cpython.buffer cimport PyBUF_FORMAT, PyBUF_ND, PyBUF_WRITABLE
+from cpython.buffer cimport PyBUF_FORMAT, PyBUF_ND, PyBUF_READ, PyBUF_WRITABLE
+from cpython.memoryview cimport (
+    PyMemoryView_FromMemory,
+    PyMemoryView_FromObject,
+    PyMemoryView_GET_BUFFER,
+)
```
Comment (pentschev) [inline ucp/_libs/ucx_api.pyx:18]: Seems like none of these new imports are used anywhere, are they necessary?

File: ucp/_libs/ucx_api_dep.pxd [hunk 2]
```diff
@@ -103,8 +103,6 @@ cdef extern from "ucp/api/ucp.h":
     int UCP_EP_PARAMS_FLAGS_NO_LOOPBACK
     int UCP_EP_PARAMS_FLAGS_CLIENT_SERVER
 
-    ctypedef ucp_conn_request* ucp_conn_request_h
```
Comment (pentschev) [inline ucp/_libs/ucx_api_dep.pxd:106]: Nice catch!

**Review by pentschev — APPROVED** [review]
This looks really great @madsbk , thanks a lot for writing this, especially the `ucx_api` test!

I have only a minor question about possibly unused imports, otherwise LGTM.

File: ucp/_libs/ucx_api.pyx (hunk 1 above)
Comment (madsbk) [reply ucp/_libs/ucx_api.pyx]: Nope, I ended up using `Array` instead of writing my own buffer handling :)

File: ucp/_libs/ucx_api_dep.pxd (hunk 2 above)
Comment (madsbk) [reply ucp/_libs/ucx_api_dep.pxd:106]: Yeah, the verbose flag really helps catching Cython errors when building, e.g. `pip install -v -e .`

**Comment by pentschev** [comment]
Thanks again @madsbk ! 😄

**Pull request merged**